edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -132,63 +132,153 @@ public async Task SetEndpoint(Endpoint newEndpoint, IList<uint> priorities)
             }
         }
 
-        void UpdatePriorities(IList<uint> priorities)
+        public async Task UpdatePriorities(IList<uint> priorities, Option<Endpoint> newEndpoint)
         {
+            Preconditions.CheckArgument(priorities.Count > 0);
+
+            if (this.closed)
+            {
+                throw new InvalidOperationException($"Endpoint executor for endpoint {this.Endpoint} is closed.");
+            }
+
             // Update priorities by merging the new ones with the existing.
             // We don't ever remove stale priorities, otherwise stored messages
             // pending for a removed priority will never get sent.
-            IList<uint> snapshot = this.priorities.Value;
-            this.priorities.Value = priorities
-                .Union(snapshot)
-                .OrderBy(p => p)
-                .ToList();
+            OrderedDictionary snapshot = this.prioritiesToFsms;
+            foreach (uint priority in priorities)
+            {
+                if (!snapshot.Contains(priority))
+                {
+                    string id = GetMessageQueueId(this.Endpoint.Id, priority);
+
+                    // Create a message queue in the store for every priority we have
+                    await this.messageStore.AddEndpoint(id);
+
+                    // Create a checkpointer and a FSM for every message queue
+                    ICheckpointer checkpointer = await this.checkpointerFactory.CreateAsync(id);
+                    EndpointExecutorFsm fsm = new EndpointExecutorFsm(this.Endpoint, checkpointer, this.config);
+
+                    // Add it to our dictionary
+                    snapshot.Add(priority, fsm);
+                }
+                else
+                {
+                    // Update the existing FSM with the new endpoint
+                    var fsm = (EndpointExecutorFsm)snapshot[priority];
+                    newEndpoint.ForEach(async e => await fsm.RunAsync(Commands.UpdateEndpoint(e)));
+                }
+            }
+
+            this.prioritiesToFsms.Value = snapshot;
+            this.lastUsedFsm = (EndpointExecutorFsm)snapshot[0];
+        }
+
+        static string GetMessageQueueId(string endpointId, uint priority)
+        {
+            if (priority == RouteFactory.DefaultPriority)
+            {
+                // We need to maintain backwards compatibility
+                // for existing sequential stores that don't
+                // have the "_Pri<x>" suffix. We use the default
+                // priority (2,000,000,000) for this, which means
+                // the store ID is just the endpoint ID.
+                return endpointId;

```

Comment:
So this means that even new routes that are set to the default priority won't have the "_Pri" suffix, correct? 
Also, if a customer explicitly set their priority to 2,000,000,000 it won't have pri either, right?

Reply:
yes for both, but how the underlying stores are handled is opaque to the customer anyways.  Short of cracking open the RockDB files, they'll never be able to tell that priority-2bn results in a different naming format.

---
In reply to: [389952225](https://github.com/Azure/iotedge/pull/2670#discussion_r389952225) [](ancestors = 389952225)